lib/spack/spack/test/url_fetch.py

```diff
@@ -358,3 +359,97 @@ def _which(*args, **kwargs):
             with Stage(fetcher, path=testpath) as stage:
                 out = stage.fetch()
             assert err_fmt.format('curl') in out
+
+
+def test_url_fetch_text_without_url(tmpdir):
+    with pytest.raises(web_util.FetchFailureError, match='URL is required'):
+        web_util.fetch_url_text(None)
+
+
+@pytest.mark.parametrize('_fetch_method', ['curl', 'urllib'])
+def test_url_fetch_text_with_invalid_url(tmpdir, _fetch_method):
+    with spack.config.override('config:url_fetch_method', _fetch_method):
+        with pytest.raises(web_util.FetchFailureError):
+            web_util.fetch_url_text('https://no-such-site/')
+
+
+def test_url_missing_curl(tmpdir, monkeypatch):
+    """Check if URL's curl is missing."""
+    url = 'https://github.com'
+    err_fmt = 'No such command {0}'
+
+    def _which(*args, **kwargs):
+        err_msg = err_fmt.format(args[0])
+        raise spack.util.executable.CommandNotFoundError(err_msg)
+
+    # Patching the 'which' symbol imported by web_util works since
+    # it is too late in import processing to patch the defining
+    # (spack.util.executable) module's symbol.
+    monkeypatch.setattr(spack.util.web, 'which', _which)
```

Comment:
No, it imports it.  So I'm monkey-arching the imported symbol.  I had problems before (noted in the comment).  I will try again and change if it works.

Reply:
Nope.  Change is still necessary. Note `spack.util.executable.which` is imported and used in the `skipif()` for `test_url_with_status_bar`.

Reply:
This is also the approach used for `test_missing_curl`.

Reply:
I did not realize that if module `foo` has `from bar import x`, that you have to monkeypatch `foo.x` rather than `bar.x`. This seems generally problematic (if a test manages to call both, then I could see someone forgetting to change both), but as you say it is an issue which predates this PR.